File: internal/handler/redirect.go
package handler

import (
	"errors"

	"github.com/andreseu/url-redirector/internal/repository"
	"github.com/gofiber/fiber/v2"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

type Handler struct {
	repo *repository.LinkRepository
}

func New(pool *pgxpool.Pool) *Handler {
	return &Handler{repo: repository.New(pool)}
}

// APIKeyMiddleware protege as rotas de admin com uma chave simples no header.
func APIKeyMiddleware(apiKey string) fiber.Handler {
	return func(c *fiber.Ctx) error {
		if c.Get("X-API-Key") != apiKey {
			return c.Status(fiber.StatusUnauthorized).JSON(fiber.Map{
				"error": "chave inválida",
			})
		}
		return c.Next()
	}
}

// GET /:slug — redireciona e incrementa cliques
func (h *Handler) Redirect(c *fiber.Ctx) error {
	slug := c.Params("slug")

	link, err := h.repo.GetBySlug(c.Context(), slug)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return c.Status(fiber.StatusNotFound).JSON(fiber.Map{
				"error": "link não encontrado",
			})
		}
		return err
	}

	// Incrementa cliques em background — não bloqueia o redirect
	go func() {
		_ = h.repo.IncrementClicks(c.Context(), slug)
	}()

	return c.Redirect(link.TargetURL, fiber.StatusMovedPermanently)
}

// POST /api/links
type createRequest struct {
	Slug      string `json:"slug"`
	TargetURL string `json:"target_url"`
}

func (h *Handler) CreateLink(c *fiber.Ctx) error {
	var body createRequest
	if err := c.BodyParser(&body); err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error": "corpo inválido",
		})
	}
	if body.Slug == "" || body.TargetURL == "" {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error": "slug e target_url são obrigatórios",
		})
	}

	link, err := h.repo.Create(c.Context(), body.Slug, body.TargetURL)
	if err != nil {
		return err
	}

	return c.Status(fiber.StatusCreated).JSON(link)
}

// GET /api/links
func (h *Handler) ListLinks(c *fiber.Ctx) error {
	links, err := h.repo.List(c.Context())
	if err != nil {
		return err
	}
	return c.JSON(links)
}

// DELETE /api/links/:slug
func (h *Handler) DeleteLink(c *fiber.Ctx) error {
	slug := c.Params("slug")
	if err := h.repo.Delete(c.Context(), slug); err != nil {
		return err
	}
	return c.SendStatus(fiber.StatusNoContent)
}

File: src/handler/redirect.ts
import { Request, Response, NextFunction } from "express";
import * as repo from "../repository/link";

const SLUG_PATTERN = /^[A-Za-z0-9_-]{1,64}$/;

function isValidHttpUrl(value: string): boolean {
  try {
    const parsed = new URL(value);
    return parsed.protocol === "http:" || parsed.protocol === "https:";
  } catch {
    return false;
  }
}

export function apiKeyMiddleware(req: Request, res: Response, next: NextFunction) {
  if (!process.env.API_KEY) {
    res.status(500).json({ error: "API_KEY não configurada" });
    return;
  }
  if (req.headers["x-api-key"] !== process.env.API_KEY) {
    res.status(401).json({ error: "Chave inválida" });
    return;
  }
  next();
}

export async function redirect(req: Request, res: Response) {
  const link = await repo.getBySlug(req.params.slug);
  if (!link) {
    res.status(404).json({ error: "Link não encontrado" });
    return;
  }
  repo.incrementClicks(req.params.slug).catch(console.error);
  res.redirect(301, link.target_url);
}

export async function createLink(req: Request, res: Response) {
  const slug = typeof req.body.slug === "string" ? req.body.slug.trim() : "";
  const targetUrl = typeof req.body.target_url === "string" ? req.body.target_url.trim() : "";

  if (!slug || !targetUrl) {
    res.status(400).json({ error: "Slug e target_url são obrigatórios" });
    return;
  }

  if (!SLUG_PATTERN.test(slug)) {
    res.status(400).json({ error: "Slug inválido. Use até 64 caracteres com letras, números, '_' ou '-'" });
    return;
  }

  if (!isValidHttpUrl(targetUrl)) {
    res.status(400).json({ error: "target_url inválida. Use uma URL http ou https completa" });
    return;
  }

  const link = await repo.create(slug, targetUrl);
  res.status(201).json(link);
}

export async function listLinks(_req: Request, res: Response) {
  const links = await repo.list();
  res.json(links);
}

export async function getStats(_req: Request, res: Response) {
  const stats = await repo.stats();
  res.json(stats);
}

export async function deleteLink(req: Request, res: Response) {
  await repo.remove(req.params.slug);
  res.sendStatus(204);
}
